### PR TITLE
add krt-utils library with KRT validation and build

### PR DESCRIPTION
**Pull request by guumaster**

### CODE REVIEW

**Review by marcvivancos — APPROVED** [review]

**Review by marcvivancos — DISMISSED** [review]

File: libs/krt-utils/pkg/builder/builder.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	if file == "" {
+		return false, nil
+	}
+
+	for _, p := range patterns {
+		if p == file {
+			return true, nil
+		}
+		match, err := zglob.Match(p, file)
+		if err != nil {
```
Comment (guumaster) [inline libs/krt-utils/pkg/builder/builder.go:70]: Maybe you can join `if`s statements into one? `if match || err != nil { return match, err }`

File: libs/krt-utils/pkg/builder/builder.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	v := validator.New()
+
+	for _, file := range ymls {
+		k, err := v.ParseFile(file)
+		if err == nil {
+			return k, nil
+		}
+	}
+
+	return nil, errors.New("no yaml file found")
```
Comment (guumaster) [inline libs/krt-utils/pkg/builder/builder.go:93]: move `errors.New(...)` to a constant.

File: libs/krt-utils/pkg/validator/validator.go [hunk 3]
```diff
@@ -218,3 +218,50 @@ func validateResourceName(fl validator.FieldLevel) bool {
 func validateEnvVar(fl validator.FieldLevel) bool {
 	return reEnvVar.MatchString(fl.Field().String())
 }
+
+//func (v *Validator) ValidateKRTFile(filepath string) (bool, error) {
```
Comment (guumaster) [inline libs/krt-utils/pkg/validator/validator.go:222]: :scream: commented out code (i know, is still a draft)

File: libs/krt-utils/pkg/builder/builder_test.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+		},
+		{
+			File: map[string]bool{
+				"konstellation-io.go":           true,
+				"/foo/bar/konstellation-io.log": true,
+			},
+			Patterns: []string{"**/konstellation-io*"},
+		},
+		{},
+		//{
```
Comment (guumaster) [inline libs/krt-utils/pkg/builder/builder_test.go:116]: remove comments when finished testing.

File: libs/krt-utils/pkg/builder/builder.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+	// produce gzip
+	if err = zr.Close(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Build validates source dir and creates a KrtFile.
```
Comment (guumaster) [inline libs/krt-utils/pkg/builder/builder.go:172]: Maybe split into two? `Validate()` and `Build()` ?

**Review by marcvivancos — APPROVED** [review]

File: libs/krt-utils/pkg/builder/builder.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+	}
+
+	data, _ := ioutil.ReadFile(ignoreFile)
+	p := strings.Split(string(data), "\n")
+	patterns := b.cleanPatterns(dir, p)
+
+	return patterns
+}
+
+// SkipFile checks with a set of clean patterns if a file has to be skipped or not.
```
Comment (guumaster) [inline libs/krt-utils/pkg/builder/builder.go:73]: this method is public or private?

Comment (marcvivancos) [reply libs/krt-utils/pkg/builder/builder.go:73]: yeah this makes much more sense as a private method...

File: libs/krt-utils/pkg/validator/errors.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+	var e error
+
+	for i := 0; i < len(ve); i++ {
+		e = ve[i]
+		buff.WriteString(e.Error())
+		buff.WriteString("\n")
+	}
+
+	return strings.TrimSpace(buff.String())
+}
```
Comment (angelsanchez) [inline libs/krt-utils/pkg/validator/errors.go:25]: why are we defining ValidationErrors as an slice of errors?
Note that the Validate function is just returning one error:
`ValidationErrors{errors.New(details)}`

File: krt-template/greeter/krt.yml [hunk 8]
```diff
@@ -4,6 +4,11 @@ entrypoint:
   proto: public_input.proto
   image: konstellation/kre-entrypoint:1.6.0
 
+config:
+  variables:
+    - KEY1
+    - KEY2
```
[the remark this answers is not in the document]
Comment (marcvivancos) [reply krt-template/greeter/krt.yml:10]: This is a mistake will fix it.

[the remark this answers is not in the document]
Comment (guumaster) [reply krt-template/greeter/krt.yml:10]: probably my mistake to test setting variables.

File: libs/krt-utils/pkg/validator/errors.go (hunk 7 above)
Comment (guumaster) [reply libs/krt-utils/pkg/validator/errors.go:25]: This is the same technique used in `github.com/go-playground/validator`, this file has a comment that says:

```
// ValidationErrors is an array of Error's to store multiple error messages post validation.
// NOTE: this technique based on github.com/go-playground/validator.
```
This way you can return an array that can be programmatically formatted separately or as a single error with a single `fmt.Print`

Comment (angelsanchez) [reply libs/krt-utils/pkg/validator/errors.go:25]: I understand the technique but as you can see in the Validate function we are sending just a long string containing all the validation errors.

**Review by angelsanchez — APPROVED** [review]

**Pull request merged**